pokemon-app/src/utils/effectiveness.js

```diff
@@ -133,14 +140,149 @@ const moveEffectiveness = (defenderType, attackerMoveTypes) => {
+  // 1 if otherwise.
+  let stab = 1;
+  attackerOverallTypes.forEach((overallTypeObject) => {
+    if (attackerMoveType === overallTypeObject.type.name) {
+      stab = 1.5;
+      return;
+    }
+  });
+
+  // The effective Attack stat of the attacker if the used move is a physical move,
+  // or the effective Special Attack stat of the attacker if the used move is a
+  // special move.
+  const attackerStats = {};
+  attackerStatsArray.forEach((statsData) => {
+    attackerStats[statsData.stat.name] = statsData.base_stat;
+  });
+  const attackerEffectiveAttackStat =
+    attackerMoveDamageClassName === "physical"
+      ? attackerStats.attack
+      : attackerMoveDamageClassName === "special"
+      ? attackerStats["special-attack"]
+      : 0;
+
+  // The effective Defense stat of the defender if the used move is a physical move
+  // or a special move that uses the defender's Defense stat, or the effective
+  // Special Defense of the defender if the used move is an other special move.
+  const defenderStats = {};
+  defenderStatsArray.forEach((statsData) => {
+    defenderStats[statsData.stat.name] = statsData.base_stat;
+  });
+  const defenderEffectiveDefendStat =
+    attackerMoveDamageClassName === "special"
+      ? defenderStats["special-defense"]
+      : defenderStats.defense;
+
+  const effectiveAttackStatRatio =
+    attackerEffectiveAttackStat / defenderEffectiveDefendStat;
+
+  // Get the move's generation for further calculations.
+  const attackerMoveGeneration = attackerMoveGenerationName.split("-").pop();
+
+  // 2 for a critical hit in Generations III-V, 1.5 for a critical hit from
+  // Generation VI onward, and 1 otherwise.
+  const critical =
+    attackerMoveGeneration === "iii" ||
+    attackerMoveGeneration === "iv" ||
+    attackerMoveGeneration === "v"
+      ? 2
+      : attackerMoveGeneration === "vi" ||
+        attackerMoveGeneration === "vii" ||
+        attackerMoveGeneration === "viii"
+      ? 1.5
+      : 1;
+
+  // Calculate the damage.
+  const damage =
+    ((((2 * attackerLevel) / 5 + 2) * power * effectiveAttackStatRatio) / 50 +
+      2) *
+    targets *
+    weather *
+    badge *
+    critical *
+    stab *
+    type *
+    burn *
+    other;
+
+  // A random factor for the damage: From Generation III onward, it is a random
+  // integer percentage between 0.85 and 1.00 (inclusive). In Generations I and II,
+  // it is realized as a multiplication by a random uniformly distributed integer
+  // between 217 and 255 (inclusive), followed by an integer division by 255.
+  const damageLowerLimit =
+    attackerMoveGeneration === "i" || attackerMoveGeneration === "ii"
+      ? (217 * damage) / 255
+      : 0.85 * damage;
+
+  const damageUpperLimit =
+    attackerMoveGeneration === "i" || attackerMoveGeneration === "ii"
+      ? (255 * damage) / 255
+      : 1 * damage;
+
+  // Return the damage, based on luck.
+  return randomGeneratorInclusive(damageLowerLimit, damageUpperLimit);
```

Comment:
this is just insane

Reply:
Can't get over how well thought out this is!